src/include/migraphx/generic_float.hpp

```diff
@@ -0,0 +1,392 @@
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include <migraphx/config.hpp>
+#include <migraphx/bit_cast.hpp>
+#include <algorithm>
+#include <limits>
+#include <iostream>
+#include <bitset>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+
+template<unsigned int N>
+constexpr unsigned int all_ones() noexcept
+{
+    return (1 << N) - 1;
+}
+
+struct float32_parts 
+{
+    unsigned int mantissa : 23;
+    unsigned int exponent : 8;
+    unsigned int sign : 1;
+
+    static constexpr unsigned int mantissa_width()
+    {
+        return 23;
+    }
+
+    static constexpr unsigned int max_exponent()
+    {
+        return all_ones<8>();
+    }
+
+    static constexpr int exponent_bias()
+    {
+        return all_ones<7>();
+    }
+
+    constexpr float to_float() const noexcept
+    {
+        return migraphx::bit_cast<float>(*this);
+    }
+};
+
+constexpr float32_parts get_parts(float f)
+{
+    return migraphx::bit_cast<float32_parts>(f);
+}
+
+
+template<unsigned int MantissaSize, unsigned int ExponentSize, unsigned int Flags = 0>
+struct __attribute__((packed)) generic_float
+{
+    unsigned int mantissa : MantissaSize;
+    unsigned int exponent : ExponentSize;
+    unsigned int sign : 1;
+
+    static constexpr int exponent_bias()
+    {
+        return all_ones<ExponentSize - 1>();
+    }
+
+    explicit constexpr generic_float(float f = 0.0) noexcept
+    {
+        from_float(get_parts(f));
+    }
+
+    constexpr generic_float &operator=(float f) noexcept
+    {
+        from_float(get_parts(f));
+        return *this;
+    }
+
+    constexpr generic_float operator-() const noexcept
+    {
+        generic_float result = *this;
+        result.sign = !this->sign;
+        return result;
+    }
+
+    constexpr float to_float() const noexcept
+    {
+        float32_parts f{};
+
+        f.sign = sign;
+
+        if(exponent == 0)
+        {
+
+            if(mantissa == 0)
+            {
+
+                f.exponent = 0;
+                f.mantissa = 0;
+            }
+            else
+            {
+                int shift  = 0;
+                f.mantissa = mantissa;
+
+                while((f.mantissa & (1 << MantissaSize)) == 0)
+                {
+                    f.mantissa <<= 1;
+                    shift++;
+                }
+
+                f.mantissa &= all_ones<MantissaSize>();
```

Comment:
Why is this needed? It look like to me it doesnt do anything. The mantissa is not going to ever be bigger than `all_ones<MantissaSize>()`.